test/proxy_test.go

```diff
@@ -0,0 +1,716 @@
+	proxyEnv := os.Getenv("HTTPS_PROXY")
+	os.Setenv("HTTPS_PROXY", pLis.Addr().String())
+	defer func() { os.Setenv("HTTPS_PROXY", proxyEnv) }()
+
+	// Use the httpproxy package functions instead of `http.ProxyFromEnvironment`
+	// because the latter reads proxy-related environment variables only once at
+	// initialization. This behavior causes issues when running multiple tests
+	// with different proxy configurations, as changes to environment variables
+	// during tests would be ignored. By using `httpproxy.FromEnvironment()`, we
+	// ensure proxy settings are read dynamically in each test execution.
+	origHTTPSProxyFromEnvironment := delegatingresolver.HTTPSProxyFromEnvironment
+	delegatingresolver.HTTPSProxyFromEnvironment = func(req *http.Request) (*url.URL, error) {
+		return httpproxy.FromEnvironment().ProxyFunc()(req.URL)
+	}
+	defer func() {
+		delegatingresolver.HTTPSProxyFromEnvironment = origHTTPSProxyFromEnvironment
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	conn, err := grpc.Dial(unresolvedTargetURI, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("grpc.Dial(%s) failed: %v", unresolvedTargetURI, err)
+	}
+	defer conn.Close()
+
+	// Send an empty RPC to the backend through the proxy.
+	client := testgrpc.NewTestServiceClient(conn)
+	if _, err := client.EmptyCall(ctx, &testgrpc.Empty{}); err != nil {
+		t.Errorf("EmptyCall failed: %v", err)
+	}
+
+	select {
+	case err := <-errCh:
+		t.Fatalf("proxy server encountered an error: %v", err)
+	case <-doneCh:
+		t.Logf("proxy server succeeded")
+	}
+}
+
+// Tests the scenario where `grpc.Dial` is performed with a proxy and the "dns"
+// scheme for the target. The test verifies that the proxy URI is correctly
+// resolved and that the target URI resolution on the client preserves the
+// original behavior of `grpc.Dial`. It also ensures that a connection is
+// established to the proxy server, with the resolved target URI sent in the
+// HTTP CONNECT request, successfully connecting to the backend server.
+func (s) TestGRPCDialWithDNSAndProxy(t *testing.T) {
+	backendAddr := createAndStartBackendServer(t)
+	pLis, errCh, doneCh, _ := setupProxy(t, backendAddr, false, requestCheck(backendAddr))
+
+	// Overwrite the proxy environment and restore it after the test.
+	proxyEnv := os.Getenv("HTTPS_PROXY")
+	os.Setenv("HTTPS_PROXY", unresolvedProxyURI)
+	defer func() { os.Setenv("HTTPS_PROXY", proxyEnv) }()
+
+	origHTTPSProxyFromEnvironment := delegatingresolver.HTTPSProxyFromEnvironment
+	delegatingresolver.HTTPSProxyFromEnvironment = func(req *http.Request) (*url.URL, error) {
+		return httpproxy.FromEnvironment().ProxyFunc()(req.URL)
+	}
+	defer func() {
+		delegatingresolver.HTTPSProxyFromEnvironment = origHTTPSProxyFromEnvironment
+	}()
+
+	// Configure manual resolvers for both proxy and target backends
+	targetResolver := setupDNS(t)
```

Comment:
We can use the real DNS resolver to resolve "localhost". Why use a fake hostname and manual resolver?

Reply:
I cannot use a local host because the `http.FromProxyEnvirnoment` as well as `httpproxy.FromEnviornment` returns a nil proxy URL if the target URI is local host. The documentation says :

[`http.FromProxyEnvirnoment`](https://pkg.go.dev/net/http#ProxyFromEnvironment)
As a special case, if req.URL.Host is "localhost" (with or without a port number), then a nil URL and nil error will be returned.

[`http.FromProxyEnvirnoment`](https://pkg.go.dev/golang.org/x/net/http/httpproxy#Config.ProxyFunc) 
As a special case, if req.URL.Host is "localhost" or a loopback address (with or without a port number), then a nil URL and nil error will be returned.